### PR TITLE
Versions query

### PR DESCRIPTION
**Pull request by mayerm94**
Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is:  https://oncprojectracking.healthit.gov/support/browse/LANTERN-320
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [ ] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [ ] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [ ] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.

### CODE REVIEW

File: capabilityreceiver/pkg/capabilityhandler/capabilityreceiver.go [hunk 1]
```diff
@@ -262,6 +267,22 @@ func saveMsgInDB(message []byte, args *map[string]interface{}) error {
...
 
+func removeNoLongerExistingVersionsInfos(ctx context.Context, store *postgresql.Store, url string, supportedVersions []string) error {
+	// If there is a requestedVersion for a URL in fhir_endpoints_info that is no longer in supportedVersions
+	// then we need to remove those fhir_endpoint_info entries
+	endptInfos, err := store.GetFHIREndpointInfosByURLWithDifferentRequestedVersion(ctx, url, supportedVersions)
+	if err != nil {
+		return err
+	}
+	for _, infoEntry := range endptInfos {
+		err = store.DeleteFHIREndpointInfo(ctx, infoEntry)
```
Comment (bmath10) [inline capabilityreceiver/pkg/capabilityhandler/capabilityreceiver.go:278]: This is just a question and I think is kind of out of scope of this ticket, but why does `DeleteFHIREndpointInfo` take a whole FhirEndpointInfo object? It seems like kind of overkill to have to get all of the fields of the table in `GetFHIREndpointInfosByURLWithDifferentRequestedVersion` when we really just need the IDs? I was just wondering why `DeleteFHIREndpointInfo` was set up that way and if there's something I'm missing

File: capabilityreceiver/pkg/capabilityhandler/capabilityreceiver_integration_test.go [hunk 2]
```diff
@@ -71,7 +71,157 @@ func TestMain(m *testing.M) {
...
+		VendorID:              fakeVendor.ID,
+		TLSVersion:            "TLS 1.1",
+		MIMETypes:             []string{"application/json+fhir"},
+		CapabilityStatement:   nil,
+		SMARTResponse:         nil,
+		RequestedFhirVersion:  "1.0.0",
+		CapabilityFhirVersion: "1.0.2",
+		Metadata:              endpointMetadataRequestVersion1}
+
+	err := setup()
```
Comment (bmath10) [inline capabilityreceiver/pkg/capabilityhandler/capabilityreceiver_integration_test.go:120]: Why is `setup()` called here and in `Test_saveMsgInDB` if it's also called in the `TestMain` function?

File: endpointmanager/pkg/endpointmanager/postgresql/fhirendpointinfostore.go [hunk 3]
```diff
@@ -462,6 +464,92 @@ func (s *Store) DeleteFHIREndpointInfo(ctx context.Context, e *endpointmanager.F
 	return err
 }
 
+// GetFHIREndpointInfosByURLWithDifferentRequestedVersion
```
Comment (bmath10) [inline endpointmanager/pkg/endpointmanager/postgresql/fhirendpointinfostore.go:467]: Should probably still have a full comment here (even if the name is pretty self explanatory)

File: endpointmanager/pkg/versionsoperatorparser/versionsresponse.go [hunk 4]
```diff
@@ -45,9 +45,13 @@ func (vr *VersionsResponse) GetDefaultVersion() string {
 
 // GetDefaultVersion gets the default FHIR version out of the versions response
```
Comment (bmath10) [inline endpointmanager/pkg/versionsoperatorparser/versionsresponse.go:46]: Copy-and-pasted comment

**Review by bmath10 — CHANGES_REQUESTED** [review]
A few more comments that I couldn't leave on specific files:

1. If we're using "None" instead of "", I think we would want to update the `dbsetup.sql` and the most recent up migration script since Emily had it so that the `requested_fhir_version` was set to ""
2. This is incredibly nitpicky, but if we're updating the metadata for every url/requested_fhir_version pair, then the TODO in `fhirendpointinfostore_test` (line 118) can be removed, since we initially thought the metadata wouldn't be added for everything.

File: capabilityreceiver/pkg/capabilityhandler/capabilityreceiver_integration_test.go (hunk 2 above)
Comment (mayerm94) [reply capabilityreceiver/pkg/capabilityhandler/capabilityreceiver_integration_test.go:120]: I think this was an artifact of my debugging. I removed it and it doesn't appear to have changed the test behavior.

File: capabilityreceiver/pkg/capabilityhandler/capabilityreceiver.go (hunk 1 above)
Comment (mayerm94) [reply capabilityreceiver/pkg/capabilityhandler/capabilityreceiver.go:278]: Good question, I don't see why there couldn't be a "DeleteFHIREndpoitInfoByID" -- But I think since everywhere we use the DeleteFHIREndpointInfo we have the entire entry the decision of whether to take in the whole entry or just the ID is 6 of one, half dozen of the other

Comment (bmath10) [reply capabilityreceiver/pkg/capabilityhandler/capabilityreceiver.go:278]: yeah that's fair, i was just curious!

**Review by bmath10 — APPROVED** [review]
